File: Modules/ServerTypes/InitYear/InitYearForm.ts
import { IntegerEditor, EnumEditor, BooleanEditor, PrefixedContext } from "@serenity-is/corelib";
import { LeaveCarryForward } from "./LeaveCarryForward";
import { AnnualLeavePolicyEditor } from "@/HumanResource/AnnualLeavePolicy/AnnualLeavePolicy/AnnualLeavePolicyEditor";
import { AnnualLeaveJobGradePolicyEditor } from "@/HumanResource/AnnualLeaveJobGradePolicy/AnnualLeaveJobGradePolicy/AnnualLeaveJobGradePolicyEditor";
import { SickLeavePolicyEditor } from "@/HumanResource/SickLeavePolicy/SickLeavePolicy/SickLeavePolicyEditor";
import { initFormType } from "@serenity-is/corelib/q";

export interface InitYearForm {
    Year: IntegerEditor;
    LeaveBringForwardMethod: EnumEditor;
    HospitalisationLeave: IntegerEditor;
    PaternityLeave: IntegerEditor;
    MaternityLeave: IntegerEditor;
    CompassionateLeave: IntegerEditor;
    MarriageLeave: IntegerEditor;
    MonthOfServiceToEligibleForMaternityLeave: IntegerEditor;
    MonthOfServiceToEligibleForPaternityLeave: IntegerEditor;
    BringForwardDays: IntegerEditor;
    BringForwardPercentage: IntegerEditor;
    LeaveRoundUp: BooleanEditor;
    PolicyList: AnnualLeavePolicyEditor;
    AnnulLeaveBasedOnJobGrade: AnnualLeaveJobGradePolicyEditor;
    SickLeavePolicyList: SickLeavePolicyEditor;
}

export class InitYearForm extends PrefixedContext {
    static formKey = 'HumanResource.InitYear.InitYear';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!InitYearForm.init)  {
            InitYearForm.init = true;

            var w0 = IntegerEditor;
            var w1 = EnumEditor;
            var w2 = BooleanEditor;
            var w3 = AnnualLeavePolicyEditor;
            var w4 = AnnualLeaveJobGradePolicyEditor;
            var w5 = SickLeavePolicyEditor;

            initFormType(InitYearForm, [
                'Year', w0,
                'LeaveBringForwardMethod', w1,
                'HospitalisationLeave', w0,
                'PaternityLeave', w0,
                'MaternityLeave', w0,
                'CompassionateLeave', w0,
                'MarriageLeave', w0,
                'MonthOfServiceToEligibleForMaternityLeave', w0,
                'MonthOfServiceToEligibleForPaternityLeave', w0,
                'BringForwardDays', w0,
                'BringForwardPercentage', w0,
                'LeaveRoundUp', w2,
                'PolicyList', w3,
                'AnnulLeaveBasedOnJobGrade', w4,
                'SickLeavePolicyList', w5
            ]);
        }
    }
}

[LeaveCarryForward]; // referenced types

File: Modules/HumanResource/InitYear/InitYear/InitYearForm.cs
using HRMSoftware.AnnualLeaveJobGradePolicy;
using HRMSoftware.AnnualLeavePolicy;
using HRMSoftware.SickLeavePolicy;
using Serenity.ComponentModel;
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace HRMSoftware.InitYear.Forms;

[FormScript("HumanResource.InitYear.InitYear")]
[BasedOnRow(typeof(InitYearRow), CheckNames = true)]
public class InitYearForm
{
    [Tab("Basic Company Information")]
    [FullWidth]
    public int Year { get; set; }
    

    // public string BasedCountry { get; set; }
    [FullWidth]
    public int LeaveBringForwardMethod { get; set; }



    [HalfWidth]
    public int HospitalisationLeave { get; set; }
    [HalfWidth]
    public int PaternityLeave { get; set; }
    [HalfWidth]
    public int MaternityLeave { get; set; }
    [HalfWidth]
    public int CompassionateLeave { get; set; }
    [FullWidth]
    public int MarriageLeave { get; set; }
    [Category("Month Of Service To Eligible For")]
    [HalfWidth]
    public int MonthOfServiceToEligibleForMaternityLeave { get; set; }
    [HalfWidth]
    public int MonthOfServiceToEligibleForPaternityLeave { get; set; }

    [OneThirdWidth]
    public int BringForwardDays { get; set; }
    [OneThirdWidth]
    public int BringForwardPercentage { get; set; }
    [OneThirdWidth]
    public int LeaveRoundUp { get; set; }


    [Category("")]
    [Tab("Annual Leave(Year of Service)"), IgnoreName, LabelWidth("0"), AnnualLeavePolicyEditor]
    public List<AnnualLeavePolicyRow> PolicyList { get; set; }

    [Category("")]
    [Tab("Annual Leave(Job Grade)"), IgnoreName, LabelWidth("0"), AnnualLeaveJobGradePolicyEditor]
    public List<AnnualLeaveJobGradePolicyRow> AnnulLeaveBasedOnJobGrade { get; set; }

    [Category("")]
    [Tab("Sick Leave Policy"), IgnoreName, LabelWidth("0"), SickLeavePolicyEditor]
    public List<SickLeavePolicyRow> SickLeavePolicyList { get; set; }


}